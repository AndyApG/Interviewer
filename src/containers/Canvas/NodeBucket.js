import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'recompose';
import Node from '../Node';
import { DragSource, DropObstacle } from '../../behaviours/DragAndDrop';
import { NO_SCROLL } from '../../behaviours/DragAndDrop/DragManager';
import { entityPrimaryKeyProperty } from '../../ducks/modules/network';
import { makeGetNextUnplacedNode } from '../../selectors/canvas';

const EnhancedNode = DragSource(Node);

class NodeBucket extends PureComponent {
  render() {
    const {
      allowPositioning,
      node,
    } = this.props;

    if (!allowPositioning || !node) { return (<div />); }

    return (
      <div className="node-bucket">
        { node
          && (
          <EnhancedNode
            key={node[entityPrimaryKeyProperty]}
            meta={() => ({ ...node, itemType: 'POSITIONED_NODE' })}
            scrollDirection={NO_SCROLL}
            {...node}
          />
          )}
      </div>
    );
  }
}

NodeBucket.propTypes = {
  allowPositioning: PropTypes.bool,
  node: PropTypes.object,
};

NodeBucket.defaultProps = {
  allowPositioning: true,
  node: null,
};

const makeMapStateToProps = () => {
  const getNextUnplacedNode = makeGetNextUnplacedNode();

  const mapStateToProps = (state, {
    layoutVariable, subject, sortOrder, stage,
  }) => {
    const node = getNextUnplacedNode(state, {
      layoutVariable, subject, sortOrder, stage,
    });

    return {
      node,
    };
  };

  return mapStateToProps;
};

export { NodeBucket };

export default compose(
  connect(makeMapStateToProps),
  DropObstacle,
)(NodeBucket);
