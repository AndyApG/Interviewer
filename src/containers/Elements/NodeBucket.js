import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { first } from 'lodash';
import { Node } from 'network-canvas-ui';
import { makeGetNextUnplacedNode, makeGetSociogramOptions, sociogramOptionsProps } from '../../selectors/sociogram';
import { draggable } from '../../behaviours';
import { actionCreators as networkActions } from '../../ducks/modules/network';

const EnhancedNode = draggable(Node);
const label = node => node.nickname;

const draggableType = 'POSITIONED_NODE';

export class NodeBucket extends Component {
  static propTypes = {
    node: PropTypes.object,
    updateNode: PropTypes.func.isRequired,
    ...sociogramOptionsProps,
  };

  static defaultProps = {
    node: null,
  };

  onDropNode = (hits, coords, node) => {
    const { layoutVariable } = this.props;

    const hit = first(hits);
    const relativeCoords = {
      x: (coords.x - hit.x) / hit.width,
      y: (coords.y - hit.y) / hit.height,
    };

    this.props.updateNode({ ...node, [layoutVariable]: relativeCoords });
  };

  render() {
    const {
      node,
    } = this.props;

    if (!node) { return null; }

    return (
      <div className="node-bucket">
        { node &&
          <EnhancedNode
            label={label(node)}
            onDropped={(hits, coords) => this.onDropNode(hits, coords, node)}
            draggableType={draggableType}
            {...node}
          />
        }
      </div>
    );
  }
}

function makeMapStateToProps() {
  const getNextUnplacedNode = makeGetNextUnplacedNode();
  const getSociogramOptions = makeGetSociogramOptions();

  return function mapStateToProps(state, props) {
    return {
      node: getNextUnplacedNode(state, props),
      ...getSociogramOptions(state, props),
    };
  };
}

function mapDispatchToProps(dispatch) {
  return {
    updateNode: bindActionCreators(networkActions.updateNode, dispatch),
  };
}

export default connect(makeMapStateToProps, mapDispatchToProps)(NodeBucket);
