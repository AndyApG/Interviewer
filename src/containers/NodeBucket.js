import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'recompose';
import { Node } from 'network-canvas-ui';
import { makeGetNextUnplacedNode, makeGetSociogramOptions } from '../selectors/sociogram';
import { DragSource } from '../behaviours/DragAndDrop';

const EnhancedNode = DragSource(Node);
const label = node => node.nickname;

class NodeBucket extends Component {
  static propTypes = {
    node: PropTypes.object,
  };

  static defaultProps = {
    node: null,
  };

  render() {
    const {
      node,
    } = this.props;

    if (!node) { return null; }

    return (
      <div className="node-bucket">
        { node &&
          <EnhancedNode
            label={label(node)}
            meta={() => ({ ...node, itemType: 'POSITIONED_NODE' })}
            {...node}
          />
        }
      </div>
    );
  }
}

function makeMapStateToProps() {
  const getNextUnplacedNode = makeGetNextUnplacedNode();
  const getSociogramOptions = makeGetSociogramOptions();

  return function mapStateToProps(state, props) {
    return {
      node: getNextUnplacedNode(state, props),
      ...getSociogramOptions(state, props),
    };
  };
}

export { NodeBucket };

export default compose(
  connect(makeMapStateToProps),
)(NodeBucket);
