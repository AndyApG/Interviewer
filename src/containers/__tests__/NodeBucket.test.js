/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import { NodeBucket } from '../NodeBucket';

const sociogramOptionsDefault = {
  layoutVariable: 'foo',
  createEdge: 'bar',
  displayEdges: [],
  canCreateEdge: false,
  canHighlight: false,
  highlightAttributes: {},
  allowSelect: false,
  allowPositioning: false,
  selectMode: '',
  nodeBinSortOrder: {},
  concentricCircles: 0,
  skewedTowardCenter: false,
};

const mockProps = {
  ...sociogramOptionsDefault,
  node: { label: 'some label' },
  updateNode: () => {},
  layout: 'foo',
  sort: {},
  getLabel: node => node.label,
};

describe('<NodeBucket />', () => {
  it('renders ok', () => {
    const component = shallow(<NodeBucket {...mockProps} />);

    expect(component).toMatchSnapshot();
  });
});
