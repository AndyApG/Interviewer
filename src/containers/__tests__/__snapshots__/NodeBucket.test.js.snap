// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NodeBucket /> renders ok 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NodeBucket
    allowPositioning={true}
    allowSelect={false}
    canCreateEdge={false}
    canHighlight={false}
    concentricCircles={0}
    createEdge="bar"
    displayEdges={Array []}
    getLabel={[Function]}
    highlightAttributes={Object {}}
    layout="foo"
    layoutVariable="foo"
    node={
      Object {
        "label": "some label",
      }
    }
    selectMode=""
    skewedTowardCenter={false}
    sort={Object {}}
    sortOrder={Array []}
    updateNode={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <DragSource
        allowDrag={true}
        label="some label"
        meta={[Function]}
        scrollDirection="NO_SCROLL"
      />,
      "className": "node-bucket",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "allowDrag": true,
        "label": "some label",
        "meta": [Function],
        "scrollDirection": "NO_SCROLL",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <DragSource
          allowDrag={true}
          label="some label"
          meta={[Function]}
          scrollDirection="NO_SCROLL"
        />,
        "className": "node-bucket",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "allowDrag": true,
          "label": "some label",
          "meta": [Function],
          "scrollDirection": "NO_SCROLL",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
